### PR TITLE
Avoid using early returns to prevent classname duplication

### PR DESCRIPTION
**Pull request by youknowriad**
## What?

As we've seen recently class_exists + early returns is not enough to prevent against classes duplication properly. It actually depends on the php version and sometimes PHP scans the classes before running the code which results in php fatal errors.

This PR updates our code to avoid these.

## Why?

We're merging more and more code into Core and this will help us prevent fatal errors.

## How?

I just inverted the condition, I wanted to update our PHPCS rule to not consider the early return as a correct fix but failed to do so. I'll leave that for @anton-vlasenko either here or as a follow-up. 

## Testing Instructions

There's no functional change in this PR, also consider checking the diff with the "whitespace" option enabled.

### CODE REVIEW

File: lib/class-wp-duotone-gutenberg.php [hunk 1]
```diff
@@ -32,10 +32,6 @@
  * @since 6.3.0
  */
 
-if ( class_exists( 'WP_Duotone_Gutenberg' ) ) {
```
Comment (youknowriad) [inline lib/class-wp-duotone-gutenberg.php:35]: These have Gutenberg prefixes so there's not need to guard. They are not meant to be included with this name in core.

**Comment by anton-vlasenko** [comment]
I'm working on this PR.

File: test/php/gutenberg-coding-standards/Gutenberg/Sniffs/CodeAnalysis/GuardedFunctionAndClassNamesSniff.php [hunk 2]
```diff
@@ -165,6 +165,13 @@ private function processClassToken( File $phpcsFile, $stackPointer ) {
 			$regexp               = sprintf( '/if\s*\(\s*!\s*class_exists\s*\(\s*(\'|")%s(\'|")/', preg_quote( $className, '/' ) );
 			$result               = preg_match( $regexp, $content );
 			if ( 1 === $result ) {
+				$incorrect_guarded_error_message = sprintf(
+					'The class "%s" is not properly guarded against redeclaration. Please ensure the entire class body is wrapped within an "if ( ! class_exists( \'%s\' )) {" statement.',
+					$className,
+					$className
+				);
+
+				$phpcsFile->addError( $incorrect_guarded_error_message, $classToken, 'ClassIncorrectlyGuardedAgainstRedeclaration' );
```
Comment (cbravobernal) [inline test/php/gutenberg-coding-standards/Gutenberg/Sniffs/CodeAnalysis/GuardedFunctionAndClassNamesSniff.php]: 🚀

**Comment by youknowriad** [comment]
I think the phpcs array alignment rule is probably buggy. It forced me to do this commit [cdba242](https://github.com/WordPress/gutenberg/pull/58500/commits/cdba242c6521c1a3a6a4bd1767a6dd8409b130b6) to make it pass but I believe that commit shouldn't be needed.

**Review by anton-vlasenko — APPROVED** [review]
I approve this PR because the code looks good to me and I've tested it. However, I've also been involved in the development, so I guess my approval doesn't count.

**Comment by youknowriad** [comment]
LGTM as well

**Review by cbravobernal — APPROVED** [review]
3 LGTMs

**Comment by hellofromtonya** [comment]
Folks, an issue is open for discussing different approaches to resolving this https://github.com/WordPress/gutenberg/issues/58467. The approach taken in this PR is Approach 1, but the discussion is leaning towards Approach 2 that moves the guard into the `load.php` to not load the class file. I invite you to join in the discussion in the issue. Thanks :)

**Comment by youknowriad** [comment]
@hellofromtonya Thanks for chiming in. I want to add more context here. We already had an issue with one of the package update PRs in this release 6.5 and since there's a Gutenberg release happening today, we wanted this issue fixed for all the current usage included on that Gutenberg release. That way, when we make newer backports in the next days (like font library), we won't have conflicts with the latest Gutenberg release.

Now, for the future, I don't mind which approach we take personally.

**Comment by cbravobernal** [comment]
Cherry picked for `Gutenberg 17.6` in: https://github.com/WordPress/gutenberg/commit/3f1a1deff8364c5b4dbddd2fc86a1619781c8e8d

**Pull request merged**